File: leetcode/medium/518-coinChange2.js
/*
You are given coins of different denominations and a total amount of money. Write a function to compute the number of combinations 
that make up that amount. You may assume that you have infinite number of each kind of coin.

Example 1:
Input: amount = 5, coins = [1, 2, 5]
Output: 4
Explanation: there are four ways to make up the amount:
5=5
5=2+2+1
5=2+1+1+1
5=1+1+1+1+1

Example 2:
Input: amount = 3, coins = [2]
Output: 0
Explanation: the amount of 3 cannot be made up just with coins of 2.

Example 3:
Input: amount = 10, coins = [10] 
Output: 1
*/

// first pass solution
var change = function(amount, coins) {
    let dict = {};
    
    function compute(currSum, index) {
        if (currSum === 0) return 1
        if (currSum < 0 || !coins[index]) return 0;
        if (`${currSum}:${index}` in dict) return dict[`${currSum}:${index}`];
        
        let sum = 0;
        
        for (let i = index; i < coins.length; i++) {
            if (currSum - coins[i] >= 0) {
                sum += compute(currSum - coins[i], i);
            }
        }
        dict[`${currSum}:${index}`] = sum;

        return dict[`${currSum}:${index}`];
    }
    compute(amount, coins.length - 1);
    
    return compute(amount, 0);
};

File: leetcode/medium/518-coinChange2.java
// first pass solution
class Solution {
    public int change(int amount, int[] coins) {
        Map<String, Integer> cache = new HashMap<>();
        
        return DFS(amount, coins, cache, 0);
    }
    
    public int DFS(int currAmount, int[] coins, Map<String, Integer> cache, int index) {
        String key = currAmount + ":" + index;
        
        if (cache.containsKey(key)) return cache.get(key);
        if (currAmount == 0) return 1;
        if (currAmount < 0 || index > coins.length - 1) return 0;
        
        int ways = 0;
        
        for (int i = index; i < coins.length; i++) {
            if (currAmount - coins[i] >= 0) {
                ways += DFS(currAmount - coins[i], coins, cache, i);
            } 
        }
        cache.put(key, ways);
        
        return cache.get(key);
    }
}

// optimized bottom up DP approach
class Solution {
    public int change(int amount, int[] coins) {
        int[][] dp = new int[coins.length + 1][amount + 1];
        dp[0][0] = 1;
        
        for (int i = 1; i < dp.length; i++) {
            for (int j = 0; j < dp[0].length; j++) {
                int curr = j - coins[i-1] < 0 ? 0 : dp[i][j - coins[i-1]];
                dp[i][j] = dp[i-1][j] + curr;
            }
        }
        return dp[coins.length][amount];
    }
}
